include/RAJA/pattern/launch/launch_core.hpp

```diff
@@ -176,25 +176,49 @@ struct LaunchParams
   Threads apply(Threads const& a) { return (threads = a); }
 };
 
-class LaunchContext
+template<bool StoreDim3 = false>
+class LaunchContextT
 {
 public:
+#if defined(RAJA_ENABLE_CUDA) || defined(RAJA_ENABLE_HIP)
+  // If StoreDim3 is true, store by value; else, don't store
+  typename std::conditional<StoreDim3, dim3, void*>::type thread_id;
+  typename std::conditional<StoreDim3, dim3, void*>::type block_dim;
+#endif
+
   // Bump style allocator used to
   // get memory from the pool
   size_t shared_mem_offset;
-
   void* shared_mem_ptr;
 
 #if defined(RAJA_SYCL_ACTIVE)
   // SGS ODR issue
   mutable ::sycl::nd_item<3>* itm;
 #endif
 
-  RAJA_HOST_DEVICE LaunchContext()
+  RAJA_HOST_DEVICE LaunchContextT()
       : shared_mem_offset(0),
         shared_mem_ptr(nullptr)
   {}
 
+#if defined(RAJA_ENABLE_CUDA) || defined(RAJA_ENABLE_HIP)
+  // Only enable this constructor if StoreDim3 is true
+  template<bool S = StoreDim3, typename std::enable_if<S, int>::type = 0>
```

Comment:
```suggestion
  template<bool S = StoreDim3, std::enable_if_t<S, int>= 0>
```

Reply:
You can use this form if you want to make it harder to inadvertently provide a value explicitly.
```suggestion
  template<bool S = StoreDim3, std::enable_if_t<S>* = nullptr>
```